clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -312,8 +321,12 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
   }
 
   bool VisitDeclRefExpr(DeclRefExpr *E) {
+    Decl* D = E->getDecl();
+    if (isKnownGoodDecl(D))
```

Comment:
There's a ton more of the CheckSYCLType stuff that likely needs to still be checked.  This is skipping anything further down the AST, or components.  I'd suggest doing this at a finer grained level and simply escaping the VLA check.

Reply:
> This is skipping anything further down the AST, or components.

This is intentional. Since this function is supposed to be declared internally (in SYCL headers), we can make sure it is correct.

> I'd suggest doing this at a finer grained level and simply escaping the VLA check.

This can be a bit tricky, because all other function only check for types, and we don't have a declaration there. #835 propagates it down the call stack, but I'm not sure if it is a good thing to do.

Reply:
Let me know if you want this to be addressed.

Reply:
I'm concerned that calls further down the the printf would be skipped.  What about something like
__spirv_ocl_printf ("asdf %d", CallSSomeVariadicFunc(1,2,3,4,5,6)); ?

Reply:
> I'm concerned that calls further down the the printf would be skipped. What about something like
> __spirv_ocl_printf ("asdf %d", CallSSomeVariadicFunc(1,2,3,4,5,6)); ?

That CallSSomeVariadicFunc is supposed to be a different expression, so it should be checked by another VisitDeclRefExpr, right?

Reply:
I'd hope so?  But I'm not super sure of what CheckSYCLType does in this regard.

Reply:
@erichkeane, if I understand correctly, particular call and its argument are verified by `VisitCallExpr`